feat(core): implement new block syntax

crisbeto opened the pull request:
Switches the syntax for blocks from `{#block}{/block}` to `@block {}` based on the feedback from the community.

Read more about the decision-making process in our blog: https://blog.angular.io/meet-angulars-new-control-flow-a02c6eee7843

The existing block types changed in the following ways:

**Conditional blocks:**
```html
<!-- Before -->
{#if cond}
  Main content
  {:else if otherCond}
    Else if content
  {:else}
    Else content
{/if}

<!-- After -->
@if (cond) {
  Main content
} @else if (otherCond) {
  Else if content
} @else {
  Else content
}
```

**Deferred blocks**
```html
<!-- Before -->
{#defer when isLoaded}
  Main content
  {:loading} Loading...
  {:placeholder} <icon>pending</icon>
  {:error} Failed to load
{/defer}

<!-- After -->
@defer (when isLoaded) {
  Main content
} @loading {
  Loading...
} @placeholder {
  <icon>pending</icon>
} @error {
  Failed to load
}
```

**Switch blocks:**
```html
<!-- Before -->
{#switch value}
  {:case 1}
    One
  {:case 2}
    Two
  {:default}
    Default
{/switch}

<!-- After -->
@switch (value) {
  @case (1) {
    One
  }

  @case (2) {
    Two
  }

  @default {
    Default
  }
}
```

**For loops**
```html
<!-- Before -->
{#for item of items; track item}
  {{item.name}}
  {:empty} No items
{/for}

<!-- After -->
@for (item of items; track item) {
  {{item.name}}
} @empty {
  No items
}
```

crisbeto commented:
Note for reviewer: the Saucelabs failure is unrelated.

AndrewKushnir commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -224,46 +224,57 @@ class _Tokenizer {
-    this._requireCharCode(chars.$RBRACE);
-    this._endToken([name]);
+    this._attemptCharCodeUntilFn(code => {
+      if (code === chars.$SPACE) {
```
We should probably consider other whitespace symbols here as well.

AndrewKushnir commented on packages/compiler/src/render3/r3_control_flow.ts:
```diff
@@ -22,44 +22,53 @@ const FOR_LOOP_TRACK_PATTERN = /^track\s+(.*)/;
+  for (const block of connectedBlocks) {
+    const children = html.visitAll(visitor, block.children, block.children);
+
+    if (block.name === 'else if') {
```
It'd be great to add a test to verify that we handle whitespace symbols (tabs, etc) between "else" and "if". We can do it in a followup PR as well.

AndrewKushnir commented on packages/compiler/src/render3/r3_template_transform.ts:
```diff
@@ -93,6 +93,7 @@ class HtmlAstToIvyAst implements html.Visitor {
   // This array will be populated if `Render3ParseOptions['collectCommentNodes']` is true
   commentNodes: t.Comment[] = [];
   private inI18nBlock: boolean = false;
+  private processedNodes = new Set<html.Block|html.Text>();
```
I'd propose adding a comments here to explain when we use this Set.

AndrewKushnir commented on packages/compiler/src/render3/r3_template_transform.ts:
```diff
@@ -317,44 +320,70 @@ class HtmlAstToIvyAst implements html.Visitor {
 
-  visitBlockGroup(group: html.BlockGroup, context: any): t.Node|null {
-    const primaryBlock = group.blocks[0];
+  visitBlock(block: html.Block, context: html.Node[]) {
+    const index = Array.isArray(context) ? context.indexOf(block) : -1;
```
Looks like we'd need to adjust the type to match it with this condition.

crisbeto replied:
The type is correct, but I have an assertion to ensure that it's correct at runtime as well.

AndrewKushnir commented on packages/compiler/src/render3/r3_template_transform.ts:
```diff
@@ -317,44 +320,70 @@ class HtmlAstToIvyAst implements html.Visitor {
-      this.reportError('Block group must have at least one block.', group.sourceSpan);
+    if (index === -1) {
+      this.reportError(
+          'Visitor invoked incorrectly. Expecting visitBlock to be invoked siblings array as its context',
```
Is it an internal error or something that developers would see? Since we do `reportError`, I guess it'd be visible to developers. Can we link it to a potential case in a template that can cause this (similar to what we had before)?

crisbeto replied:
It's an internal error for ourselves to ensure that we pass in the `context` since we don't have proper type checking for it. I switched it to a regular error since it's something we can't recover from.

AndrewKushnir commented on packages/compiler/src/render3/r3_template_transform.ts:
```diff
@@ -317,44 +320,70 @@ class HtmlAstToIvyAst implements html.Visitor {
+      return null;
+    }
+
+    if (!this.options.enabledBlockTypes.has(block.name)) {
+      let errorMessage: string;
+
+      if (CONNECTED_DEFERRED_BLOCK_TYPES.has(block.name)) {
```
I'm a bit confused by the `if (!this.options.enabledBlockTypes.has(block.name))` check - should we just throw an "unrecognized block" error in this case? It looks like the rest of the checks are needed when a block is enabled.

crisbeto replied:
The idea behind these checks is to have a better error message if we encounter a connected block in a random place, e.g. `@placeholder {}` without `@defer {}`. Normally it's the "main" blocks that are specified in the `enabledBlockTypes`, not the connected ones. Also worth noting that we have an "Unrecognized block" error in the `switch` below as well.

AndrewKushnir reviewed (approved):
@crisbeto awesome work! 👍 

I've left a few comments, but none of them are blocking. We can address them in a followup PR.

crisbeto committed:
feat(core): implement new block syntax

Switches the syntax for blocks from `{#block}{/block}` to `@block {}` based on the feedback from the community.

Read more about the decision-making process in our blog: https://blog.angular.io/meet-angulars-new-control-flow-a02c6eee7843

The existing block types changed in the following ways:

**Conditional blocks:**
```html
<!-- Before -->
{#if cond}
  Main content
  {:else if otherCond}
    Else if content
  {:else}
    Else content
{/if}

<!-- After -->
@if (cond) {
  Main content
} @else if (otherCond) {
  Else if content
} @else {
  Else content
}
```

**Deferred blocks**
```html
<!-- Before -->
{#defer when isLoaded}
  Main content
  {:loading} Loading...
  {:placeholder} <icon>pending</icon>
  {:error} Failed to load
{/defer}

<!-- After -->
@defer (when isLoaded) {
  Main content
} @loading {
  Loading...
} @placeholder {
  <icon>pending</icon>
} @error {
  Failed to load
}
```

**Switch blocks:**
```html
<!-- Before -->
{#switch value}
  {:case 1}
    One
  {:case 2}
    Two
  {:default}
    Default
{/switch}

<!-- After -->
@switch (value) {
  @case (1) {
    One
  }

  @case (2) {
    Two
  }

  @default {
    Default
  }
}
```

**For loops**
```html
<!-- Before -->
{#for item of items; track item}
  {{item.name}}
  {:empty} No items
{/for}

<!-- After -->
@for (item of items; track item) {
  {{item.name}}
} @empty {
  No items
}
```

dylhunn commented:
This PR was merged into the repository by commit 8be2c48b7cda5e99f3d1efa9f26eb2615fea6a8b.
